File: Solutions/Python/best_time_to_buy_and_sell_stock_iii.py
"""
    [HARD]
    123. Best Time to Buy and Sell Stock III

    Concepts:
    - dynamic programming
    - array

    Stats:
        Runtime | 1888 ms   [Beats 9.10%]
        Memory  | 456.84 MB [Beats 5.73%]
"""

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        """
            Memoization
        """
        n = len(prices)
        dp = defaultdict(int)
        is_buying = True

        def recurse(ind, is_buying, trs):

            # if no more transactions or
            # end of prices reached, return 0
            if not trs or ind >= n:
                return 0

            if (ind, is_buying, trs) in dp:
                return dp[(ind, is_buying, trs)]
            

            # if buying...
            if is_buying:
                dp[(ind, is_buying, trs)] = max(
                    recurse(ind+1, not is_buying, trs) - prices[ind],   # buy at current price
                    recurse(ind+1, is_buying, trs)                      # or continue
                )

            # if selling...
            else:
                dp[(ind, is_buying, trs)] = max(
                    recurse(ind+1, not is_buying, trs-1) + prices[ind], # sell at current price
                    recurse(ind+1, is_buying, trs))                     # or continue


            return dp[(ind, is_buying, trs)]

        return recurse(0, is_buying, 2)
                    

File: Solutions/Cpp/best_time_to_buy_and_sell_stock_iii.cpp
/*
    [HARD]
    123. Best Time to Buy and Sell Stock III

    Concepts:
    - array
    - dynamic programming

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 79.25 MB  [Beats 87.43%]
*/

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        const long INF = std::numeric_limits<long>::max();

        // initialize variables for first buy, first sell,
        // second buy, and second cell
        long buy1 = INF, buy2 = INF;
        long sell1 = 0, sell2 = 0;

        // iterate over prices to update buy and sell values
        for (const long price : prices) {
            // update first buy and sell values
            buy1 = std::min(buy1, price);
            sell1 = std::max(sell1, price - buy1);
            // update second buy and sell values
            buy2 = std::min(buy2, price - sell1);
            sell2 = std::max(sell2, price - buy2);
        }

        return (int)sell2;
    }
};